servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ProxyTunnel.java

```diff
@@ -105,72 +108,71 @@ int connectCount() {
             }
             if (b != '\r') {
-                bytes[i++] = (byte) b;
+                bos.write((byte) b);
             }
         }
-        return new String(bytes, 0, i, UTF_8);
+        return bos.toString(UTF_8.name());
     }
 
-    private void handleRequest(final Socket socket, final InputStream in, final String initialLine) throws IOException,
-            ExecutionException, InterruptedException {
-        try {
-            if (initialLine.startsWith(CONNECT_PREFIX)) {
-                final int end = initialLine.indexOf(' ', CONNECT_PREFIX.length());
-                final String authority = initialLine.substring(CONNECT_PREFIX.length(), end);
-                final String protocol = initialLine.substring(end + 1);
-                final int colon = authority.indexOf(':');
-                final String host = authority.substring(0, colon);
-                final int port = Integer.parseInt(authority.substring(colon + 1));
-
-                try (Socket clientSocket = new Socket(host, port)) {
-                    connectCount.incrementAndGet();
-                    final OutputStream out = socket.getOutputStream();
-                    out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
-                    out.flush();
-
-                    final InputStream cin = clientSocket.getInputStream();
-                    Future<Void> f = executor.submit(() -> {
-                        copyStream(out, cin);
-                        return null;
-                    });
-                    copyStream(clientSocket.getOutputStream(), in);
-                    f.get(); // wait for the copy of proxy client input to server output to finish copying.
+    private void handleRequest(final Socket socket, final String initialLine) throws IOException, ExecutionException,
+            InterruptedException {
+        if (initialLine.startsWith(CONNECT_PREFIX)) {
+            final int end = initialLine.indexOf(' ', CONNECT_PREFIX.length());
+            final String authority = initialLine.substring(CONNECT_PREFIX.length(), end);
+            final String protocol = initialLine.substring(end + 1);
+            final int colon = authority.indexOf(':');
+            final String host = authority.substring(0, colon);
+            final int port = Integer.parseInt(authority.substring(colon + 1));
+
+            try (Socket clientSocket = new Socket(host, port)) {
+                connectCount.incrementAndGet();
+                final OutputStream out = socket.getOutputStream();
+                out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
+                out.flush();
+
+                Future<Void> f = executor.submit(() -> {
+                    try {
+                        copyStream(out, clientSocket.getInputStream());
+                    } finally {
+                        socket.shutdownOutput();
+                        clientSocket.shutdownInput();
```

Comment:
Tests are designed to verify ST behaves correctly when a legacy proxy that does not support duplex channels is used. Please, revert to the `close()` method instead.

Reply:
You can leave a comments inside this ProxyTunnel for future:
This proxy is intentionally does not support duplex channels to let `GracefulConnectionClosureHandlingTest` and `ConnectionCloseHeaderHandlingTest` verify we behave correctly with such proxies.